pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go

```diff
@@ -25,39 +27,182 @@ import (
+)
+
+func GenerateMatchMap(pctx networking.EnvoyFilter_PatchContext, efw *model.EnvoyFilterWrapper) (cpw map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	serviceMap map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	subsetMap map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	portMap map[string][]*model.EnvoyFilterConfigPatchWrapper) {
+
+	cpw = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
+	serviceMap = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
+	subsetMap = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
+	portMap = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
 	if efw == nil {
-		return
+		return cpw, serviceMap, subsetMap, portMap
 	}
 	for _, cp := range efw.Patches[networking.EnvoyFilter_CLUSTER] {
-		if cp.Operation != networking.EnvoyFilter_Patch_MERGE {
+		if cp.Operation != networking.EnvoyFilter_Patch_REMOVE &&
+			cp.Operation != networking.EnvoyFilter_Patch_MERGE {
+			continue
+		}
+		if !commonConditionMatch(pctx, cp) {
 			continue
 		}
-		if commonConditionMatch(pctx, cp) && clusterMatch(c, cp) {
+
+		if cp.Match.GetCluster() == nil {
+			if cp.Operation == networking.EnvoyFilter_Patch_REMOVE {
+				cpw[RemovePatch] = append(cpw[RemovePatch], cp)
+				continue
+			}
+			cpw[MergeAny] = append(cpw[MergeAny], cp)
+			continue
+		}
+		if cp.Match.GetCluster().Name != "" {
+			if cp.Operation == networking.EnvoyFilter_Patch_REMOVE {
+				cpw[cp.Match.GetCluster().Name+RemovePatch] = append(cpw[cp.Match.GetCluster().Name+RemovePatch], cp)
+				continue
+			}
+			cpw[cp.Match.GetCluster().Name+MergeOne] = append(cpw[cp.Match.GetCluster().Name+MergeOne], cp)
+			continue
+		}
+
+		service := cp.Match.GetCluster().Service
+		subset := cp.Match.GetCluster().Subset
+		port := cp.Match.GetCluster().PortNumber
+
+		if service != "" {
+			serviceMap[service] = append(serviceMap[service], cp)
+		} else {
+			serviceMap[MergeAny] = append(serviceMap[MergeAny], cp)
+		}
+
+		if subset != "" {
+			subsetMap[subset] = append(subsetMap[subset], cp)
+		} else {
+			subsetMap[MergeAny] = append(subsetMap[MergeAny], cp)
+		}
+
+		if port != 0 {
+			portMap[strconv.Itoa(int(port))] = append(portMap[strconv.Itoa(int(port))], cp)
+		} else {
+			portMap[MergeAny] = append(portMap[MergeAny], cp)
+		}
+	}
+	return cpw, serviceMap, subsetMap, portMap
+}
+
+func ApplyClusterMergeOrRemove(c *cluster.Cluster, cpw map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	serviceMap map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	subsetMap map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	portMap map[string][]*model.EnvoyFilterConfigPatchWrapper) (out *cluster.Cluster, shouldKeep bool) {
+	defer runtime.HandleCrash(func(interface{}) {
+		log.Errorf("clusters patch caused panic, so the patches did not take effect")
+	})
+
+	shouldKeep = true
+	if len(cpw[RemovePatch]) > 0 || len(cpw[c.Name+RemovePatch]) > 0 {
+		c = nil
+		shouldKeep = false
+		return nil, shouldKeep
+	}
+	if len(cpw[MergeAny]) > 0 {
```

Comment:
same here

Reply:
when there are envoyFilters do not specify anything and their patchOperation is merge